### ISSUE TITLE
arangodump failed : "unable to find context (it could be expired)"

### ISSUE THREAD

**Issue by avive99** [arangodb/arangodb#4654]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [v ] 3.3

Mode:
- [v ] Single-Server

Storage-Engine:
- [v ] rocksdb

On this operating system:
- [ ] Linux 
  - [ v] RedHat .rpm

Trying to dump a db with two collections, each one with more than 200M documents.
Dump failed after few minutes, created just a few small (total 39mb) files.

The output of arangodump is :
```
> arangodump --output-directory "/tmp/avi/dump/" --server.database bezeqcloudcollectordb --overwrite true
Please specify a password:
Server version: 3.3.3
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/tmp/avi/dump'
Last tick provided by server is: 285439233
# Dumping document collection 'files'...
2018-02-22T12:51:22Z [10882] ERROR got error from server: HTTP 404 (Not Found): ArangoError 400: replication dump - unable to find context (it could be expired)
Processed 1 collection(s), wrote 39846362 byte(s) into datafiles, sent 4 batch(es)
```
The contents of output dir :
```
> ls  -ltrh
total 39M
-rw------- 1 arangodb arangodb 516 Feb 22 14:44 files_45b963397aa40d4a0063e0d85e4fe7a1.structure.json
-rw------- 1 arangodb arangodb  70 Feb 22 14:44 dump.json
-rw------- 1 arangodb arangodb 39M Feb 22 14:51 files_45b963397aa40d4a0063e0d85e4fe7a1.data.json
```
The first dumped ("files") collection properties:
```
127.0.0.1:8529@bezeqcloudcollectordb> db.files.count()
281974888
127.0.0.1:8529@bezeqcloudcollectordb> db.files.properties()
{
  "isSystem" : false,
  "waitForSync" : false,
  "keyOptions" : {
    "type" : "traditional",
    "allowUserKeys" : true,
    "lastValue" : 282137756
  },
  "cacheEnabled" : false
}
127.0.0.1:8529@bezeqcloudcollectordb> db.files.figures()
{
  "indexes" : {
    "count" : 1,
    "size" : 2392473663
  },
  "documentsSize" : 49181099001
}
```
What's the problem with arangodump in that case ?

regards

**Comment by avive99** [comment]
Made few more trials.

1. Ran arangodump with no-data option "--dump-data false" - worked.

2. Ran arangodump for another , very small db with a few records in each collection - worked.

3. The named db has just two collections.
One named "users", has 223K documents.
Ran arangodump just for that "users" collection - worked.

4. The second collection, named "files", has about 280M documents (and it's data seems to catch about 49GB on disk).
Ran arangodump just for that "files" collection - fails with the same error messages :

Are there any db/collection size limitations in ArangoDB, which could impact arangodump ?

**Comment by avive99** [comment]
The error message in the one-collection-dump failure was :
arangodump --output-directory "/tmp/avi/dump/bezeqcloudcollectordb" --server.database bezeqcloudcollectordb --collection files --overwrite true
Please specify a password:
Server version: 3.3.3
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/tmp/avi/dump/bezeqcloudcollectordb'
Last tick provided by server is: 285815985
# Dumping document collection 'files'...
2018-02-28T09:27:03Z [15450] ERROR got invalid response from server: Request timeout reached
Processed 1 collection(s), wrote 110625618 byte(s) into datafiles, sent 6 batch(es)

**Comment by avive99** [comment]
Tried to run arangoexport for that large collection, and this failed also :

arangoexport --output-directory "/tmp/avi/dump/bezeqcloudcollectordb" --server.database bezeqcloudcollectordb --collection files  --overwrite true
Please specify a password:
Connected to ArangoDB 'http+tcp://127.0.0.1:8529', version 3.3.3, database: 'bezeqcloudcollectordb', username: 'root'
# Exporting collection 'files'...
{"code":500,"error":true,"errorMessage":"out of memory (while executing) (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/RestHandler/RestCursorHandler.cpp:134)","errorNum":3}
2018-02-28T10:10:32Z [17807] ERROR {startup} startup aborted: caught exception during start of feature 'Export': got invalid response from server: HTTP 500: Internal Server Error (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangosh/Export/ExportFeature.cpp:577). Please report this error to arangodb.com. shutting down
2018-02-28T10:10:32Z [17807] ERROR {startup} arangoexport terminated because of an unhandled exception: startup aborted: caught exception during start of feature 'Export': got invalid response from server: HTTP 500: Internal Server Error (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangosh/Export/ExportFeature.cpp:577). Please report this error to arangodb.com (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/lib/ApplicationFeatures/ApplicationServer.cpp:638). Please report this error to arangodb.com

**Comment by dothebart** [comment]
Hi, 
it seems that the context controlling the dump is deleted because of the client coming back to it to late. 
can you try to specify smaller sizes here so that frequency goes up: 

  --batch-size <uint64>                              maximum size for individual data batches (in bytes) (default: 67108864)
  --initial-batch-size <uint64>                      initial size for individual data batches (in bytes) (default: 8388608)

**Comment by avive99** [comment]
Hey @dothebart ,

  Many thanks for the reply.

  Just to note that the host has 16GB of RAM and 20GB of swap size set.

  As per your advises, i've ran arangodump only for the large "files" collection few more times :

1. first time, added only the --batch-size of 10MB.
This failed:
> arangodump --batch-size 10485760 --output-directory "/tmp/avi/dump/bezeqcloudcollectordb" --server.database bezeqcloudcollectordb --collection files --overwrite true
Please specify a password:
Server version: 3.3.3
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/tmp/avi/dump/bezeqcloudcollectordb'
Last tick provided by server is: 285875758
# Dumping document collection 'files'...
2018-03-01T08:32:32Z [32619] ERROR got error from server: HTTP 404 (Not Found): ArangoError 400: replication dump - unable to find context (it could be expired)
Processed 1 collection(s), wrote 8388617 byte(s) into datafiles, sent 2 batch(es)

2. Second time, i used --batch-size of 10MB and --initial-batch-size of 1MB.
This also failed :
> arangodump --batch-size 10485760 --initial-batch-size 1048576  --output-directory "/data/avidump" --server.database bezeqcloudcollectordb --collection files --overwrite true
Server version: 3.3.3
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/data/avidump'
Last tick provided by server is: 285879490
# Dumping document collection 'files'...
2018-03-01T09:58:50Z [3236] ERROR got error from server: HTTP 500 (Internal Server Error): ArangoError 3: out of memory (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/lib/Basics/StringBuffer.h:263)
Processed 1 collection(s), wrote 74221810 byte(s) into datafiles, sent 12 batch(es)

3. Third time, i used --batch-size of 10MB and --initial-batch-size of 100KB.
This also failed :
> arangodump --batch-size 10485760 --initial-batch-size 102400 --output-directory "/data/avidump" --server.database bezeqcloudcollectordb --collection files --overwrite true
(bad: failed after 12 mins)
Server version: 3.3.3
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/data/avidump'
Last tick provided by server is: 285879982
# Dumping document collection 'files'...
2018-03-01T10:17:26Z [3959] ERROR got error from server: HTTP 404 (Not Found): ArangoError 400: replication dump - unable to find context (it could be expired)
Processed 1 collection(s), wrote 284564131 byte(s) into datafiles, sent 37 batch(es)


4. Finally i've a combination which seems to work without failures : --batch-size of 1MB and --initial-batch-size of 100KB.
But this works very slowly. The throughput of arangodump is currently about 150MB per hour. So, for a collection of almost 50GB - it shall work 14 days ?!

Any advises of how to fasten the process ?

Best regards and looking forward your assistance,
Avi

**Comment by dothebart** [comment]
We have actually fixed quiet a bunch of things in that area with the upcomming 3.3.4 release; Please hold on it will be ready before your last dump attempt is finished ;-P

**Comment by dothebart** [comment]
If the QA goes well, 3.3.4 will be out in the next days - why I expressed the hope that it'd be faster than the slow dumping.

**Comment by dothebart** [comment]
Ok, 3.3.4 is available for download. Can you upgrade and retry?

**Comment by OmarAyo** [comment]
Hi @avive99 

Did you upgrade to 3.3.4 ? - Could you please give us an update on this ticket?

**Comment by avive99** [comment]
Hey @OmarAyo 

I've upgraded to 3.3.4.
Yesterday i've run the aragondump.
(All the databases together at that host take about 49GB disk storage. There is about 52GB free on the dump destination disk.)
I've the the arangodump on 21-,mar at 15:03 as :
arangodump --server.password mypass --output-directory "/data/avidump" --server.database bezeqcloudcollectordb --collection files --overwrite true 

The dump ran during 16.5 hours, produced 52GB of data (how could it be more than the original 49G of db data ?), and finally failed this morning at 07:31 - since no more space left at the destination disk :
Server version: 3.3.4
Connected to ArangoDB 'tcp://127.0.0.1:8529', database: 'bezeqcloudcollectordb', username: 'root'
Writing dump to output directory '/data/avidump'
Last tick provided by server is: 287218062
# Dumping document collection 'files'...
2018-03-22T05:31:02Z [21579] ERROR cannot write: No space left on device
2018-03-22T05:31:02Z [21579] ERROR cannot write to file '/data/avidump/files_45b963397aa40d4a0063e0d85e4fe7a1.data.json'
Processed 1 collection(s), wrote 55472187082 byte(s) into datafiles, sent 831 batch(es)

So, there is some progress: running the arangodump without any parameters didn't fail after few minutes (as in 3.3.3), but rather continued to work.

However, few questions still remain:
Q1. Is there a way to speed up the arangodump work ? It worked 16.5 hours to dump 49GB of data, and that's a lot of time.
Q2. Could it be that arangodump isn't intended to dump large quantities of data ? May be arangodump was designed to dump just rather small databases till (let's say) 1GB - and this also could take between half an hour till a hour...
Q3. Why does arangodump produces json file larger that the original databases' size ? (52+G > 49G)

best regards and looking forward your assistance

**Comment by jsteemann** [comment]
The data produced by arangodump is uncompressed JSON, which may be bigger space-wise than the actual data stored in the database. This is because the database will store data in VelocyPack format, which normally has a space advantage compared to JSON, and the RocksDB engine will also use compression (Snappy compression by default) for most of the higher LSM levels. So chances are the database data are actually a bit smaller than its equivalent in uncompressed JSON.
Additionally, arangodump will write some extra information out with each document in a dump, e.g. "type" and "tick". This also makes a dump a bit more verbose than the plain documents' JSON data.

We are working on a branch to make arangodump dump collections in parallel by the way. This will speed up dumps when multiple collections are dumped.

**Comment by avive99** [comment]
Hey @jsteemann,

Many thanks for the fast and detailed (as usual) answer :-)
So, the current throughput of arangodump in my case is as expected, and the speedup will be available only in future releases (once parallel is ready), right ?

**Comment by jsteemann** [comment]
The parallel dump branch is currently under development. When it is ready, it will be merged into devel, which will eventually become a stable 3.4.
The changes are quite big, so we cannot simply backport them to stable 3.3. 
A speedup should be observable once there are multiple collections to dump from in parallel. The modification in that branch will not speed up dumping a single collection. But normally there are multiple collections to dump, and often several of them are big. Here we expect a real speedup.

Additionally we have also been experimenting with a few format modifications for the dump API and other APIs, but so far we haven't achieved a breakthrough there.

**Comment by jsteemann** [comment]
@avive99 : following up on this...
We just merged a change into our devel branch that will make arangodump process multiple collections at a time, using multiple threads. This may speed up dumps considerably, provided the dump is taken not just for a single (huge) collection but multiple ones of about the same size. That is at least the ideal use case. But making the dump parallel may also help in cases when collection sizes are not equally distributed.
The parallel dump feature will be release with ArangoDB 3.4.

**Comment by dothebart** [comment]
@avive99 if you want to give the parallel dump a try, you can download the nightlies; We also offer a client package, which you could install in a chroot or another host.

**Comment by vinaypyati** [comment]
Hi @avive99,

Now that we have 3.4 RC 1 released, did you have a chance to try the parallel dump feature? You can find the ArangoDB 3.4 RC 1 [here ](https://www.arangodb.com/download-technical-preview/). Let us know your feedback!

Thanks!

**Comment by maxkernbach** [comment]
Hi @avive99,

did you already try 3.4 RC1?
FYI: The latest available 3.4 Release Candidate is RC2 and can be found in the [download section](https://www.arangodb.com/download-technical-preview/).

Your feedback is highly appreciated!

**Comment by vinaypyati** [comment]
Hi @avive99,

We haven't received feedback from you for a while, I am closing this issue for now.

Please consider testing this issue on 3.4.0 RC3 which includes parallel dump feature. You can download it from the [download section](https://www.arangodb.com/download-technical-preview/). If it does not resolve your problem, please feel free to reopen the issue by adding a new comment.

Thanks!
